Enhance SearchFilter.QueryType and FailedFuture

kevinhinterlong opened the pull request:
This adds some minor enhancements to FailedFuture and allows for a QueryType to be found from its serialized name.

kevinhinterlong committed:
Add ability to find QueryType from it's serialized string.

kevinhinterlong committed:
Add a completedFuture which returns or throws as soon as it's called

- This replaces the FailedFuture by being able to be return a value or throw an exception

kevinhinterlong committed:
Delete FailedFuture

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SearchFilter.java:
```diff
@@ -42,6 +42,20 @@
+        /**
+         * Get the QueryType enum fromType it's search type.
+         * @param type  Type of the query type (for serialization)
+         * @return the enum QueryType
```
Empty lines before `@return`

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SearchFilter.java:
```diff
@@ -42,6 +42,20 @@
+
+        /**
+         * Get the QueryType enum fromType it's search type.
+         * @param type  Type of the query type (for serialization)
```
Empty lines before `@param`

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/util/CompletedFuture.java:
```diff
@@ -0,0 +1,60 @@
+// Copyright 2016 Yahoo Inc.
```
2017

kevinhinterlong committed:
Address comments

kevinhinterlong committed:
Change Copyright year

asifmansoora reviewed (approved):
This PR looks good to me 👍

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SearchFilter.java:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2016 Yahoo Inc.
+// Copyright 2017 Yahoo Inc.
```
Looks like this file is there since 2016 (we put based on when file is initialized)

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SearchFilter.java:
```diff
@@ -42,6 +42,22 @@
+         *
+         * @return the enum QueryType
+         */
+        public static QueryType fromType(String type) {
```
There is a standard way(simpler and more readable) of achieving your purpose. Example will be https://github.com/yahoo/fili/blob/master/fili-core/src/main/java/com/yahoo/bard/webservice/config/CacheFeatureFlag.java or http://grepcode.com/file_/repo1.maven.org/maven2/javax.ws.rs/jsr311-api/1.1/javax/ws/rs/core/Response.java/?v=source

Instead of `fromType`, you have 
```java
        public String getType() {
            return type;
        }
```
And you check type by having "contains".equals(QueryType.Contains.getType()). That's how we check type of enum (as the two link example does, as well)

kevinhinterlong replied:
I'm not sure I understand what you mean. Do you want me to enumerate each possibility to find which type the enum is instead of using the loop?

From the second example you gave they did this which is nearly identical to the way I checked the type.

```java
public static Status fromStatusCode(final int statusCode) {
    for (Status s : Status.values()) {
        if (s.code == statusCode) {
            return s;
        }
    }
    return null;
}
```

kevinhinterlong replied:
What's the reason for adding `final`? 
I've been told in previous PRs to remove `final` from parameters

kevinhinterlong committed:
Add changelog entry showing changes. Fix year

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/util/CompletedFuture.java:
```diff
@@ -0,0 +1,60 @@
+            throw new ExecutionException(throwable);
+        } else {
+            return item;
+        }
```
`else` is not necessary

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/util/CompletedFuture.java:
```diff
@@ -0,0 +1,60 @@
+
+    /**
+     * Creates a completed future with an item to return or
+     * if a nonnull throwable is passed it it will throw an
```
passed it -> provided ?

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/util/CompletedFuture.java:
```diff
@@ -0,0 +1,60 @@
+    /**
+     * Creates a completed future with an item to return or
+     * if a nonnull throwable is passed it it will throw an
+     * execution exception.
```
`{@link java.util.concurrent.ExecutionException}` makes it more readable

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/util/CompletedFuture.java:
```diff
@@ -0,0 +1,60 @@
+    private final Throwable throwable;
+
+    /**
+     * Creates a completed future with an item to return or
```
`an item to return on on {@link #get()}`

kevinhinterlong replied:
I think this is clear from the context and adding this could make it unclear what happens when `{@link #get(long, TimeUnit)}` is called.

QubitPi commented on fili-core/src/test/java/com/yahoo/bard/webservice/models/druid/client/impl/TestDruidWebService.java:
```diff
@@ -90,7 +90,7 @@ public TestDruidWebService() {
      * If TestDruidWebService#throwable is set, invokes the failure callback.  Otherwise invokes success or failure
      * dependent on whether TestDruidWebService#statusCode equals 200.
-     * Note that since this doesn't send requests to druid all the responses will be null or {@link FailedFuture}.
+     * Note that since this doesn't send requests to druid all the responses will be null or {@link CompletedFuture}.
```
Since this class is in a different package

`{@link com.yahoo.bard.webservice.util.CompletedFuture}`

kevinhinterlong replied:
It's available from context because of 
```java
import com.yahoo.bard.webservice.util.CompletedFuture;
```


do you want me to add the fully qualified name?

QubitPi replied:
This is not for readability. The reason to put complete path is for Javadoc plugin to be able to recognize and inject HTML link so that points to the correct class.

kevinhinterlong replied:
The javadoc plugin is able to correctly recognize this and inject the html link in other classes I've tried (like in CompletedFuture). Did you generate the javadoc and it didn't work correctly?

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/util/CompletedFuture.java:
```diff
@@ -0,0 +1,60 @@
+     * @param item  The item to return when called.
+     * @param throwable  A non null throwable if an exception should be thrown.
+     */
+    public CompletedFuture(E item, Throwable throwable) {
```
I see all constructor is called everywhere with `item = null`, like `new CompletedFuture<>(null, throwable)`. This makes code less readable, because people don't know that `null` means. A better approach as I learned from Rick's suggestion is

```java
public CompletedFuture(E item, Throwable throwable) {
    this.item = item;
    this.throwable = throwable;
}

public CompletedFuture(Throwable throwable) {
    this(null, throwable);
}
```

That way, you do `new CompletedFuture<>(throwable)`.

kevinhinterlong replied:
I've decided to replace this with [Either](https://github.com/yahoo/fili/blob/master/fili-core/src/main/java/com/yahoo/bard/webservice/util/Either.java) and give one constructor for the item and one for the throwable so it's clear you can only do one or the other.

kevinhinterlong commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SearchFilter.java:
```diff
@@ -42,6 +42,22 @@
         }
+
+        /**
+         * Get the QueryType enum from it's search type.
```
I think this is clear from context as you can only pass in a `String` to the method.

kevinhinterlong commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SearchFilter.java:
```diff
@@ -42,6 +42,22 @@
+        /**
+         * Get the QueryType enum from it's search type.
+         *
+         * @param type  The type of query (i.e. "insensitive_contains")
```
Same as above

kevinhinterlong committed:
Addressed comments

- Clarify usage of CompletedFuture by limiting how it can be created
- add two helper methods to SearchFilter

kevinhinterlong committed:
Only use utility functions and java core CompletableFuture

archolewa commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SearchFilter.java:
```diff
@@ -42,6 +42,22 @@
+         *
+         * @return the enum QueryType.
+         */
+        public static QueryType fromType(String type) {
```
I would argue that this should be returning an `Optional` instead of a `null`, or throwing an `IllegalArgumentException` if the `type` is not a recognized `QueryType`. 

Returning an Optional makes it clear to people using the code that they may not get a QueryType back, and need to deal with that case. This makes sense if it's totally valid to have a string type that cannot be translated to a QueryType. 

The second case assumes that it's an error if we get an unrecognized type.

I'm assuming that it's an error to receive an unrecognized type, but I don't know yet what it's meant to be used for, so I can't say for certain.

kevinhinterlong replied:
@QubitPi Are you fine with me reverting to the [IllegalArgumentException](https://github.com/yahoo/fili/pull/396/commits/0ced60b2512fc2eb621603e92d84bac4ca110395#r123132356)/returning optional?

For context, this was made because `SearchFilter` doesn't have a reference to its `QueryType` and I have to process a `SearchFilter` differently for each case (and if not found, throw an exception).

Additionally, if `SearchFilter` must be one of those `QueryType` then it should probably be verified in the [constructor](https://github.com/yahoo/fili/blob/369787a86097e19c672072c8fe4e5b76e4fc012a/fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SearchFilter.java#L67) that accepts a `String` queryType.

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/util/CompletedFuture.java:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * Creates a completed {@link java.util.concurrent.Future} which will return this item.
+     *
+     * @param item  The item to return when called.
```
no period

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/util/CompletedFuture.java:
```diff
@@ -0,0 +1,47 @@
+     * Creates a completed {@link java.util.concurrent.Future} which will return this item.
+     *
+     * @param item  The item to return when called.
+     * @param <E>  The type to be returned by this future.
```
no period

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/util/CompletedFuture.java:
```diff
@@ -0,0 +1,47 @@
+     * @param item  The item to return when called.
+     * @param <E>  The type to be returned by this future.
+     *
+     * @return a completed {@link CompletableFuture} which will successfully return an item.
```
no period and all others as well

kevinhinterlong committed:
address comments

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/util/CompletedFuture.java:
```diff
@@ -0,0 +1,47 @@
+     *
+     * @return a completed {@link CompletableFuture} which will successfully return an item.
+     */
+    public static <E> CompletableFuture<E> returning(E item) {
```
Where is this method used?

kevinhinterlong replied:
I was using this in #430 but it's no longer needed and can be deleted if that makes sense.

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/util/CompletedFuture.java:
```diff
@@ -0,0 +1,47 @@
+     * @param item  The item to return when called
+     * @param <E>  The type to be returned by this future
+     *
+     * @return a completed {@link CompletableFuture} which will successfully return an item.
```
no period at end

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SearchFilter.java:
```diff
@@ -42,6 +43,22 @@
+        /**
+         * Get the QueryType enum from it's search type.
+         *
+         * @param type  The type of query (i.e. "insensitive_contains")
```
`(i.e. "insensitive_contains")` indicates this is the only QueryType? I would remove it.

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SearchFilter.java:
```diff
@@ -78,18 +95,26 @@ private SearchFilter(Dimension dimension, String type, String value) {
     }
 
+    private String getQueryType() {
+        return query.get("type");
```
`type` is used multiple times. It's more maintainable to have it as a static final String.

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SearchFilter.java:
```diff
@@ -78,18 +95,26 @@ private SearchFilter(Dimension dimension, String type, String value) {
+    }
+
+    private String getQueryValue() {
+        return query.get("value");
```
same

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/util/CompletedFuture.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * A utility class to create futures which have already been completed
+ * and will return as soon as it is called.
+ *
```
extra line

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/util/CompletedFuture.java:
```diff
@@ -0,0 +1,47 @@
+
+/**
+ * A utility class to create futures which have already been completed
+ * and will return as soon as it is called.
```
can be move to the previous line

QubitPi commented on fili-core/src/main/java/com/yahoo/bard/webservice/util/CompletedFuture.java:
```diff
@@ -0,0 +1,47 @@
+     */
+    public static <E> CompletableFuture<E> throwing(Throwable throwable) {
+        CompletableFuture<E> completedFuture = CompletableFuture.supplyAsync(() -> null);
+        completedFuture.completeExceptionally(throwable);
```
@michael-mclawhorn @garyluoex @archolewa I'm little concerned about this statement. According to its [doc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#completeExceptionally-java.lang.Throwable-)

"If **not already completed**, causes invocations of get() and related methods to throw the given exception."

Looks like this method wants to throw exception **on complete** instead (Or Javadoc of this method is simply wrong?).

kevinhinterlong committed:
Partially address comments

kevinhinterlong committed:
Finish addressing comments
